docs/refman/lprsrv.8.pod
#
# mouse:~ppr/src/docs/refman/lprsrv.8.pod
# Copyright 1995--2006, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 10 April 2006.
#

=head1 NAME

lprsrv - PPR RFC 1179 (lpr/lpd) server

lprsrv.conf - Access control file

lprsrv-test - Access control test program

=head1 SYNOPSIS

B<lprsrv> I<options>

B<lprsrv-test> I<hostname> [I<username>] ...


=head1 DESCRIPTION

The B<--help> option prints a brief description of the options.
The B<--version> option prints the PPR version information.

B<lprsrv> is an RFC-1179 (Berkeley lpr/lpd) compatible print server.  It
will accept print jobs over the network and hand them to B<ppr> for 
processing.

=head2 Disabling Other RFC 1179 Servers

If you want to run B<lprsrv> on the standard port (and you almost certainly do),
you must disable any RFC 1179 server already running on that port.  This
section contains instructions for disabling various RFC 1179 servers. 
If your system's print server isn't covered in this section and you figure
out how to disable it, please send the necessary instructions to
ppr-bugs@trincoll.edu.

=head3 Berkely LPD

If you are using B<lpd> from Berkely (which is popular on Linux),
you must either run B<lprsrv> on a port other than 515, or you must arrange
for B<inetd> to be started before B<lpd>.  If B<inetd> is started before B<lpd>,
B<lpd> will be unable to bind to port 515 but will still work in other respects
and B<lprsrv> will pass remote jobs to B<lpr>.  (Unfortunately, some versions
of B<lpd> will die when they fail to bind to the port.  Can anyone give
better advice than this?)

=head3 LPR-NG

If you are using LPR-NG, you must add a line such as this to F</etc/lpd.conf>:

 lpd_port=1501

You can choose any unused port that you wish.  If you wish you can also restrict
access to the loopback device:

 lpd_port=127.0.0.1%1501

You should then restart lpd.

=head3 IRIX

If you are using IRIX 6.3, you should edit F</etc/init.d/lp> and find the place
where F</usr/lib/lpsched> is started.  Add the option "-nobsd" to the line.
That will make sure LP doesn't grab the port at the next reboot.  So that you
don't have to reboot after making the change, run these commands:

 # /usr/lib/lpshut
 # /usr/lib/lpsched -nobsd

=head3 SunOS 5.6 (Solaris 2.6)

If you are using SunOS 5.6 (Solaris 2.6) then you need only edit /etc/inetd.conf
and remove the line which reads:

    printer stream tcp nowait root /usr/lib/print/in.lpd in.lpd

Before this change will take effect, it is necessary to tell B<inetd> to reload
F<inetd.conf> by sending it the HUP signal.  If B<inetd> has the process id 123,
then the command for reloading F<inetd.conf> is:

    kill -HUP 123

or, if your B<kill> command doesn't support the modern syntax:

    kill -1 123

=head3 CUPS

[Information missing.]


=head2 Recommended Setup

By default B<pprd> will listen on port 515 and launch B<lprsrv> whenever a
new connexion comes in.  This feature will be enabled only if port 515 is
free and B<pprd> is started as root.


=head2 Alternative Setup With Inetd

If you are hacking PPR, you may want to be able to start and stop B<pprd> 
without having root access.  In that case, you should configure 
B<inetd> to launch B<lprsrv> as user B<ppr>.  The configuration 
line to insert in F</etc/inetd.conf>.

    printer stream tcp nowait ppr /usr/lib/ppr/lprsrv lprsrv


=head2 Queue Listing Options

When a client requests a queue listing for a PPR queue, B<lprsrv> will
return the output of B<ppop lpq> for the short format and B<ppop list> for
the long format.  The B<ppop lpq> output is designed to look similiar to the
output of BSD B<lpq>.

If the B<-A> switch is used, B<lprsrv> will pass the switch and its argument
on to B<ppop>.  The argument is an integer which indicates an age in
seconds.  Arrested jobs older than the specified age will not be shown in
queue listings.

When a client requests a queue listing for an LPR/LPD queue, B<lprsrv>
simply executes B<lpq> and returns the output.

When a client requests a queue listing for a System V B<lp> queue, B<lprsrv>
runs B<lpstat>.  Some B<lpq> options are not supported for B<lp> queues.


=head2 Access Control

Access to queues through B<lprsrv> is controled by the file
F</etc/ppr/lprsrv.conf>.  The only documentation currently available is in
F</etc/ppr/lprsrv.conf.sample>.

If you use an unmodified F<lprsrv.conf.sample> for F<lprsrv.conf>, then
tranditional BSD-style access control will be in effect.  Under this style
of access control, any computer listed in F</etc/hosts.equiv> or
F</etc/hosts.lpd> will be allowed to submit jobs.  The names listed in these
files should be fully qualified names.

Since B<lprsrv> uses its own code, it may not support all of the access
control syntax described in L<hosts.equiv>.  In particular, "+" and "-" are
not supported.  Domain wildcards such as ".trincoll.edu" are supported.  NIS
netgroup support is available on some systems.

If you would like to add support for additional access control syntax in
F</etc/hosts.equiv> and F</etc/hosts.lpd>, you should modify the function
authorized_file_check() in lprsrv/lprsrv_conf.c.  If you make improvements,
please consider contributing them to the PPR project.


=head2 Access Control Testing

The program B<lprsrv-test> can be used to test the rules in F</etc/ppr/lprsrv.conf>,
F</etc/hosts.equiv>, and F</etc/hosts.lpd>.

You can use it to find out what sort of access certain users on a certain
host will have.  When you run it the first parameter should be a
hostname.  The second and subsequent parameters should be usernames.

Since several sections of F</etc/ppr/lprsrv.conf> may be combined to
determine the access settings for a certain host, it can be difficult to
determine exactly what access a particular host has.  B<lprsrv-test> solves
this by printing a merged record in F</etc/ppr/lprsrv.conf> format which
shows the final settings which apply to the host you named on the command
line.

For each remote user you name on the command line, B<lprsrv-test> will print
the name of the local user which will be used to execute spooler commands on
the remote user's behalf.  The local user used may be different for each
spooler.


=head2 Other Options

=over 4

=item B<-A> I<seconds>

=item B<--arrest-interest-interval> I<seconds>

Whenever B<lprsrv> lists a PPR queue, the argument to either of these
switches is passed to the B<ppop> command as an argument for its B<-A> switch.

=back


=head2 Support for DEC RFC-1179 Extensions

The version of B<lpr> supplied with DEC OSF/1 has many more options than the
BSD version.  Using these extra options results in extra lines in the queue
file which B<lpr> sends to B<lprsrv>.  These extensions are supported by
B<lprsrv> whenever possible.


=head2 Support for Solaris RFC-1179 Extensions

The print spooler in Solaris 2.6 (SunOS 5.6) is derived from the System V
LP spooler.  When it sends jobs using the RFC-1179 protocol it include extra
control file lines which represent System V B<lp> options.  These
extensions are supported by B<lprsrv> whenever possible.


=head2 Using Switchsets with B<lprsrv>

The B<lprsrv> configuration file does not provided any way for setting
options for individual queues.  It is suggested that you use the B<ppad switchset>,
B<ppad group switchset> and B<ppad alias switchset> commands do do this.  See
the B<ppad switchset> description for examples.

Be aware that B<lprsrv> will sometimes override switchset options.  It will always
override an B<-m> option (which shouldn't be in a switchset anyway).  It will also
override duplex, paper source, and other options if conflicting instructions
are conveyed by means of Solaris or DEC extensions to the RFC-1179 protocol.


=head2 Banner Pages

The RFC 1179 protocol provides a way for a client to ask to have banner pages
suppressed.  (The BSD B<lpr -h> option uses this feature.)  If B<lprsrv>
receives a job that requests suppression of the banner page, it will attempt
to do so.  However, suppressing the banner page may not be allowed.  In that
case, the spooler to which B<lprsrv> submits the job will likely print a banner
page anyway.

In the case of a PPR printer, if the B<ppad flags> command has been used to
set the printer's banner option to B<always>, then it will not be possible for
an RFC 1179 client to suppress the banner page.

In the case of a System V B<lp> queue, an RFC 1179 client will probably not
be able to suppress the banner page unless the B<lpadmin -o> option has
been used to set a list of printer options which includes "nobanner".


=head2 Client Configuration

This section describes the commands needed to set up various print spoolers
as clients of B<lprsrv>.  If you have information to add to this section,
please send it to ppr-bugs@trincoll.edu.

=head3 BSD lpr

Add an entry similar to the following to F</etc/printcap>.

    myprn|My Printer:\
	:rp=somehost.mycoll.edu:rm=myprn:\
	:sd=/var/spool/lpd/myprn:\
	:mx#0:

Then create the local spool directory.

    mkdir /var/spool/lpd/myprn

=head1 SEE ALSO

See the comments in F</etc/ppr/lprsrv.conf>.

=head1 HISTORY

PPR was written at Trinity College during 1993--2005.

=head1 AUTHORS

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut



History.txt
mouse:~ppr/src/History.txt

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of
reading and setting terminal characteristics instead of the ioctl() method.


lprsrv/lprsrv.c
/*
** mouse:~ppr/src/lprsrv/lprsrv.c
** Copyright 1995--2006, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 23 February 2006.
*/

/*
** Berkeley LPR compatible server for PPR and LP on System V Unix.
** There is also partial support for passing jobs to LPR.
*/

#include "config.h"
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <sys/utsname.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "rfc1179.h"
#include "lprsrv.h"
#include "util_exits.h"
#include "version.h"

/*
** This function returns the name of this computer.  This is used in some
** error messages sent to the client.
*/
const char *this_node(void)
	{
	static const char *p = NULL;
	if(!p)
		{
		struct utsname u;
		uname(&u);
		p = gu_strdup(u.nodename);
		}
	return p;
	} /* end of this_node() */

/*
** This function writes a line to the lprsrv log file.  It performs 
** printf()-style formatting.
*/
static void lprsrv_vlog(const char category[], const char format[], va_list va)
	{
	FILE *logfile;
	if((logfile = fopen(LPRSRV_LOGFILE, "a")))
		{
		fprintf(logfile, "%s: %s: %ld: ", category, datestamp(), (long)getpid());
		vfprintf(logfile, format, va);
		fputc('\n', logfile);
		fclose(logfile);
		}
	} /* lprsrv_vlog() */

static void lprsrv_log(const char category[], const char format[], ...)
	{
	va_list va;
	va_start(va, format);
	lprsrv_vlog(category, format, va);
	va_end(va);
	} /* lprsrv_log() */

/*
** Print a debug line in the lprsrv log file.
*/
void debug(const char message[], ...)
	{
	va_list va;
	va_start(va, message);
	lprsrv_vlog("DEBUG", message, va);
	va_end(va);
	} /* end of debug() */

/*
** Print a warning line in the lprsrv log file.
*/
void warning(const char message[], ...)
	{
	va_list va;
	va_start(va, message);
	lprsrv_vlog("WARNING", message, va);
	va_end(va);
	} /* end of warning() */

void uprint_error_callback(const char *format, ...)
	{
	va_list va;
	va_start(va, format);
	lprsrv_vlog("UPRINT", format, va);
	va_end(va);
	} /* end of uprint_error_callback() */

/*=============================================================================
** main() and its support routines:
=============================================================================*/

/*
** Command line options:
*/
static const char *option_chars = "s:A:";
static const struct gu_getopt_opt option_words[] =
		{
		{"arrest-interest-interval", 'A', TRUE},
		{"help", 1000, FALSE},
		{"version", 1001, FALSE},
		{(char*)NULL, 0, FALSE}
		} ;

/*
** Print how to use.  The argument will be either stdout or stderr.
*/
static void help(FILE *outfile)
	{
	fputs(_("Valid switches:\n"), outfile);

	fputs(_("\t-A <seconds>\n"
		"\t--arrest-interest-interval <seconds>\n"
		"\t\t(this switch is passed thru to ppop)\n"
		"\t--version\n"
		"\t\t(print PPR version number)\n"
		"\t--help\n"
		"\t\t(print this message)\n"), outfile);
	}

/*
** main server loop,
** dispatch commands
*/
static int real_main(int argc,char *argv[])
	{
	char client_dns_name[MAX_HOSTNAME+1];
	char client_ip[16];
	int client_port;
	struct ACCESS_INFO access_info;

	/* Initialize internation messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	/*
	** Parse the command line options.  We use the parsing routine
	** in libppr.a.  All of the parsing state is kept in the
	** structure getopt_state.
	*/
	{
	int optchar;
	struct gu_getopt_state getopt_state;

	gu_getopt_init(&getopt_state, argc, argv, option_chars, option_words);

	while((optchar=ppr_getopt(&getopt_state)) != -1)
		{
		switch(optchar)
			{
			case 'A':					/* -A or --arrest-interest-interval */
				uprint_arrest_interest_interval = getopt_state.optarg;
				break;

			case 1000:					/* --help */
				help(stdout);
				exit(EXIT_OK);

			case 1001:					/* --version */
				puts(VERSION);
				puts(COPYRIGHT);
				puts(AUTHOR);
				exit(EXIT_OK);

			case '?':					/* help or unrecognized switch */
				fprintf(stderr, _("Unrecognized switch: %s\n"), getopt_state.name);
				help(stderr);
				exit(EXIT_SYNTAX);

			case ':':					/* argument required */
				fprintf(stderr, _("The %s option requires an argument.\n"), getopt_state.name);
				exit(EXIT_SYNTAX);

			case '!':					/* bad aggreation */
				fprintf(stderr, _("Switches, such as %s, which take an argument must stand alone.\n"), getopt_state.name);
				exit(EXIT_SYNTAX);

			case '-':					/* spurious argument */
				fprintf(stderr, _("The %s switch does not take an argument.\n"), getopt_state.name);
				exit(EXIT_SYNTAX);

			default:					/* missing case */
				fprintf(stderr, "Internal error: missing case for -%c option\n", optchar);
				exit(EXIT_INTERNAL);
				break;
			}
		}
	} /* end of command line parsing context */

	DODEBUG_MAIN(("connexion received"));

	/*
	** This must be done before request processing starts.  In practice it is
	** not necessary, but in theory Inetd's only guarantee is that stdin will
	**	be connected to the socket.
	*/
	dup2(0, 1);
	dup2(0, 2);

	/*
	 * Change to a known directory, set a umask (since it is difficult to
	 * know what Inetd will give us, set PPR-related variables in the 
	 * environment, and delete junk from the environment.
	 */
	chdir(LIBDIR);
	umask(PPR_UMASK);
	set_ppr_env();
	prune_env();

	/* Determine the IP address and port of the client. */
	get_client_info(client_dns_name, client_ip, &client_port);
	DODEBUG_MAIN(("connexion is from %s (%s), port %d", client_dns_name, client_ip, client_port));

	/* Search lprsrv.conf and find the information that
	   applies to this node. */
	get_access_settings(&access_info, client_dns_name);

	/* If that information indicates that the client is
	   not allowed to connect, then print and error
	   message and drop the connexion now. */
	if(! access_info.allow)
		gu_Throw(_("Node \"%s\" is not allowed to connect"), client_dns_name);
	if(! access_info.insecure_ports && client_port > 1024)
		gu_Throw(_("Node \"%s\" is not allowed to connect from insecure ports"), client_dns_name);

	/* Do zero or one commands and exit. */
	{
	char line[256];
	if(fgets(line, sizeof(line), stdin))
		{
		line[strcspn(line, "\n\r")] = '\0';

		switch(line[0])
			{
			case 1:									/* ^A */
				DODEBUG_MAIN(("start printer: \"^A%s\"", line+1));
				/* Nothing to do? */
				break;
			case 2:									/* ^B */
				DODEBUG_MAIN(("receive job: \"^B%s\"", line+1));
				do_request_take_job(line+1, client_dns_name, &access_info);
				break;
			case 3:									/* ^C */
				DODEBUG_MAIN(("short queue: \"^C%s\"", line+1));
				do_request_lpq(line);				/* This never returns */
				break;
			case 4:									/* ^D */
				DODEBUG_MAIN(("long queue: \"^D%s\"", line+1));
				do_request_lpq(line);				/* This never returns */
				break;
			case 5:										/* ^E, remove jobs */
				DODEBUG_MAIN(("remove: \"^E%s\"", line+1));
				do_request_lprm(line, client_dns_name, &access_info);
				break;
			case 0:
				gu_Throw("empty command");
			default:								/* what can we do? */
				if(line[0] < ' ')
					gu_Throw("unrecognized command: \"^%c%s\"", line[0]+'@', line+1);
				else
					gu_Throw("unrecognized command: \"%s\"", line);
			}
		} /* end of if fgets() worked */
	} /* end of line reading context */

	return 0;
	} /* end of real_main() */

int main(int argc, char *argv[])
	{
	gu_Try {
		return real_main(argc, argv);
		}
	gu_Catch {
		printf("%s\n", gu_exception);
		lprsrv_log("FATAL", "%s", gu_exception);
		exit(1);
		}
	/* NOREACHED */
	return 255;
	} /* end of main() */

/* end of file */

